Cleanup + basic HA

Sander-agfa opened the pull request:
- Added basic HA Awareness,
only deploy to active HA cluster members
- Cleanup code a bit, centralise all config in dehydrated-bigip-loadconfig
- Cleanup code, centralise functions in dehydrated-bigip-common. Remove duplicate code
- Simplify installation. Removed some files. Added file dehydrated-bigip-hookscript. Here, all combinations of DNS and deployment are possible
These changes mean that script files themselves don't have to be touched anymore. Only central configfile and hookfile need to be modified. Makes it easier to update script in case of updates

Sander-agfa committed:
Cleanup + basic HA

- Added basic HA Awareness,
only deploy to active HA cluster members
- Cleanup code a bit, centralise all config in dehydrated-bigip-loadconfig
- Cleanup code, centralise functions in dehydrated-bigip-common. Remove duplicate code
- Simplify installation. Removed some files. Added file dehydrated-bigip-hookscript. Here, all combinations of DNS and deployment are possible
These changes mean that script files themselves don't have to be touched anymore. Only central configfile and hookfile need to be modified. Makes it easier to update script in case of updates

Sander-agfa committed:
Update dehydrated-bigip-hookscript

fix filename error

Sander-agfa committed:
Fix bug in cleaning up challenges

a . (dot) and - (minus sign) can also be part of the challenge response, add this in the regex

Sander-agfa committed:
Added conditional redeploy

Added a check that allows a redeploy,
only if the remote cert is not the same as the local one.
This enables the redeploy to be cron scheduled as well, without needless re-deployment

Sander-agfa committed:
Fix to allow private key changes

Allow the case where the private key has changed, and timestamping the cert names is disabled.
Do add the timestamp here one-time

hildebrau commented on dehydrated-bigip-common:
```diff
@@ -68,6 +69,21 @@ whenAvailable() {
+isActive() {
+  OUT=$(restCall "GET" "/tm/cm/failover-status")
+  log "isActive():`echo $OUT`"
+  if [[ $(echo $OUT | jsonq '["entries"]["https://localhost/mgmt/tm/cm/failover-status/0"]["nestedStats"]["entries"]["status"]["description"]') == ACTIVE ]]; then
```
shouldn't localhost be $BIGIP_DEVICE or something?
```suggestion
  if [[ $(echo $OUT | jsonq '["entries"]["https://$BIGIP_DEVICE/mgmt/tm/cm/failover-status/0"]["nestedStats"]["entries"]["status"]["description"]') == ACTIVE ]]; then
```

hildebrau commented on dehydrated-bigip-redeploy-traffic-certificate:
```diff
@@ -1,15 +1,32 @@
+	  CERTFILE="${3}"
+	  CLIENTSSL_NAME="${1}"
+	  #Use Openssl to check the serial of the local cert, use sed to strip off "serial=" and convert to lowercase
+	  LOCALHASH=$(openssl x509 -in $CERTFILE -serial -noout |sed -E -s 's/serial=\(.*\)/\L\1/')
```
I had to remove these escape characters to get this to parse properly on my CentOS 7 box
```suggestion
	  LOCALHASH=$(openssl x509 -in $CERTFILE -serial -noout |sed -E -s 's/serial=(.*)/\L\1/')
```

Sander-agfa replied:
This alteration does not work on by ubuntu box.
Will replace with this:

```suggestion
	  LOCALHASH=$(openssl x509 -in $CERTFILE -serial -noout |sed -e 's/serial=//' | tr '[:upper:]' '[:lower:]'
```

Sander-agfa replied:
fix with latest commit in #10 .
Uses tr in stead of sed

hildebrau commented on dehydrated-bigip-common:
```diff
@@ -243,6 +263,50 @@ createClientSSLProfile() {
+		return 3
+	  fi
+    else
+      log "SSL cert link not found, return empty"
```
@Sander-agfa , what is supposed to happen if the certReference doesn't exist in the current cert?  It just seems to exit with return code 2 here and I'm not sure how to remedy the situation.

Sander-agfa replied:
@hildebrau , I should expect the function to return code 2 to the caller, and not let the script exit entirely.
I'll test this further on my dev setup, though I had this tested already.

Sander-agfa committed:
fix for sed compatibility

fix compatibility issues on sed with multiple distro's, use tr instead to convert to lowercase

Sander-agfa committed:
Add seperate config entry for data group partition

Allows the data group to be present in a different partition than the SSL profiles and virtual servers

colin-stubbs reviewed (approved):
looks good
